### PR TITLE
[Writing Tools] Rare WebContent crash in IPC::ArgumentCoder<WebKit::CoreIPCData, void>::encode

### PR DESCRIPTION
**Pull request by rr-codes**
#### ff3ae8f473dd3cf7046fd64d4349172ae57bb6a9
<pre>
[Writing Tools] Rare WebContent crash in IPC::ArgumentCoder&lt;WebKit::CoreIPCData, void&gt;::encode
<a href="https://bugs.webkit.org/show_bug.cgi?id=276289">https://bugs.webkit.org/show_bug.cgi?id=276289</a>
<a href="https://rdar.apple.com/130595277">rdar://130595277</a>

Reviewed by Aditya Keerthi.

This crash can happen when attempting to create an NSTextAttachment with a `nil` file wrapper.
Fix by instead just returning an NSAttributedString comprised of a single space in this case (to
maintain the same length as if there were an attachment).

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(attributedStringWithAttachmentForFileWrapper):
(attributedStringWithAttachmentForElement):

Canonical link: <a href="https://commits.webkit.org/280714@main">https://commits.webkit.org/280714@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/9c81eba63a92ee499987dc9e9c3e9b2ea6a20d98

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57413 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36741 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9888 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61035 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7858 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59541 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44365 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8046 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/46494 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5561 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59443 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34474 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49587 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27358 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31256 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6896 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6861 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7167 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62714 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1326 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7259 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/53752 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1332 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49619 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/53841 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1132 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8570 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32570 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33655 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34740 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33401 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 1]
```diff
@@ -2412,7 +2412,18 @@ static String preferredFilenameForElement(const HTMLImageElement& element)
     return wrapper;
 }
 
-static RetainPtr<NSAttributedString> attributedStringWithAttachmentForElement(const HTMLImageElement& element)
+template<typename Element>
+static RetainPtr<NSAttributedString> attributedStringWithAttachmentForElement(const Element& element)
+{
+    RetainPtr fileWrapper = fileWrapperForElement(element);
+    if (!fileWrapper)
+        return [[NSAttributedString alloc] initWithString:@" "];
```
Comment (whsieh) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: This is a leak. We should do one of the following:

```suggestion
        return adoptNS([[NSAttributedString alloc] initWithString:@" "]).autorelease();
```

or:

```suggestion
        return [[[NSAttributedString alloc] initWithString:@" "] autorelease];
```

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 2]
```diff
@@ -2412,7 +2412,18 @@ static String preferredFilenameForElement(const HTMLImageElement& element)
     return wrapper;
 }
 
-static RetainPtr<NSAttributedString> attributedStringWithAttachmentForElement(const HTMLImageElement& element)
+template<typename Element>
```
Comment (pxlcoder) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: Why is `template<typename Element>` needed?

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 3]
```diff
@@ -2412,7 +2412,18 @@ static String preferredFilenameForElement(const HTMLImageElement& element)
...
+static RetainPtr<NSAttributedString> attributedStringWithAttachmentForElement(const Element& element)
```
Comment (pxlcoder) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: I'd like this method name to be a bit more specific / refer to file wrappers. As it stands it's very easy for new callers to use this for `HTMLImageElement`, without calling into `attributedStringWithAttachmentForImageElement`.

File: Source/WebCore/editing/cocoa/HTMLConverter.mm [hunk 4]
```diff
@@ -2427,33 +2438,30 @@ static String preferredFilenameForElement(const HTMLImageElement& element)
...
-{
-    RetainPtr fileWrapper = fileWrapperForElement(element);
-    RetainPtr attachment = adoptNS([[PlatformNSTextAttachment alloc] initWithFileWrapper:fileWrapper.get()]);
-    return [NSAttributedString attributedStringWithAttachment:attachment.get()];
+    return attributedStringWithAttachmentForElement(element);
 }
 
 #if ENABLE(WRITING_TOOLS)
-static bool hasAncestorQualifyingForWritingToolsPreservation(Element* ancestor, WeakHashMap<Element, bool, WeakPtrImplWithEventTargetData>& cache)
+static bool hasAncestorQualifyingForWritingToolsPreservation(const Element* ancestor, WeakHashMap<Element, bool, WeakPtrImplWithEventTargetData>& cache)
```
Comment (pxlcoder) [inline Source/WebCore/editing/cocoa/HTMLConverter.mm]: Also not a big fan of an entirely unrelated change being made here...

File: Source/WebCore/editing/cocoa/HTMLConverter.mm (hunk 2 above)
Comment (rr-codes) [reply Source/WebCore/editing/cocoa/HTMLConverter.mm]: because it can accept both an HTMLImageElement or an HTMLAttachmentElement

Comment (pxlcoder) [reply Source/WebCore/editing/cocoa/HTMLConverter.mm]: Both are `Element` subclasses though.

Comment (rr-codes) [reply Source/WebCore/editing/cocoa/HTMLConverter.mm]: that won't be sufficient, because of the call to `fileWrapperForElement`, which is an overloaded method which requires either of `HTMLImageElement` or `HTMLAttachmentElement`. So `attributedStringWithAttachmentForElement` needs specifically either of those types, and not any subclass of Element

Comment (pxlcoder) [reply Source/WebCore/editing/cocoa/HTMLConverter.mm]: I see... slightly confusing to use `typename Element` given we have a very commonly used class called `Element` – but I guess it's fine...

File: Source/WebCore/editing/cocoa/HTMLConverter.mm (hunk 4 above)
Comment (rr-codes) [reply Source/WebCore/editing/cocoa/HTMLConverter.mm]: undone

File: Source/WebCore/editing/cocoa/HTMLConverter.mm (hunk 3 above)
Comment (rr-codes) [reply Source/WebCore/editing/cocoa/HTMLConverter.mm]: refactored

File: Source/WebCore/editing/cocoa/HTMLConverter.mm (hunk 2 above)
Comment (rr-codes) [reply Source/WebCore/editing/cocoa/HTMLConverter.mm]: refactored

Comment (whsieh) [reply Source/WebCore/editing/cocoa/HTMLConverter.mm]: Yeah, I _think_ we generally use something like `ElementType` here to avoid confusion.

**Review by pxlcoder — APPROVED** [review]
